Sources/Core/AWSSDKIdentity/Sources/AWSSDKIdentity/AWSCredentialIdentityResolvers/DefaultAWSCredentialIdentityResolverChain.swift

```diff
@@ -24,21 +26,57 @@ import protocol SmithyIdentity.AWSCredentialIdentityResolvedByCRT
 /// 5. EC2 Instance Metadata (IMDSv2)
 ///
 /// The credentials retrieved from the chain are cached for 15 minutes.
-public struct DefaultAWSCredentialIdentityResolverChain: AWSCredentialIdentityResolvedByCRT {
-    public let crtAWSCredentialIdentityResolver: AwsCommonRuntimeKit.CredentialsProvider
-
+public actor DefaultAWSCredentialIdentityResolverChain: AWSCredentialIdentityResolver {
     /// Creates a credential identity resolver that uses the default AWS credential identity resolver chain used by most AWS SDKs.
-    public init() throws {
-        let fileBasedConfig = try CRTFileBasedConfiguration()
-        try self.init(fileBasedConfig: fileBasedConfig)
+    typealias ResolverFactory = @Sendable () throws -> (any AWSCredentialIdentityResolver)
+
+    private let resolverFactories: [ResolverFactory]
+    private var cachedCredentials: AWSCredentialIdentity?
+
+    public init() {
+        resolverFactories = [
+            { return (try EnvironmentAWSCredentialIdentityResolver()) },
+            { return (try ProfileAWSCredentialIdentityResolver()) },
+            { return (try STSWebIdentityAWSCredentialIdentityResolver()) },
+            { return (try ECSAWSCredentialIdentityResolver()) },
+            { return (try IMDSAWSCredentialIdentityResolver()) }
+        ]
     }
 
-    @_spi(DefaultAWSCredentialIdentityResolverChain)
-    public init(fileBasedConfig: CRTFileBasedConfiguration) throws {
-        self.crtAWSCredentialIdentityResolver = try AwsCommonRuntimeKit.CredentialsProvider(source: .defaultChain(
-            bootstrap: SDKDefaultIO.shared.clientBootstrap,
-            fileBasedConfiguration: fileBasedConfig
-        ))
+    public func getIdentity(identityProperties: Attributes?) async throws -> AWSCredentialIdentity {
+        if let cached = cachedCredentials, !shouldRefreshCredentials(expiration: cached.expiration) {
+            return cached
+        }
+
+        let lastIndex = resolverFactories.count - 1
+        for index in 0..<lastIndex {
+            do {
+                let resolver = try resolverFactories[index]()
+                let credentials = try await resolver.getIdentity(identityProperties: identityProperties)
+                cachedCredentials = credentials
+                return credentials
+            } catch {
+                // Continue to the next resolver factory.
+            }
+        }
+
+        // The error thrown from the last resolver is not caught and instead gets thrown to caller.
+        let lastResolver = try resolverFactories[lastIndex]()
+        let credentials = try await lastResolver.getIdentity(identityProperties: identityProperties)
+        cachedCredentials = credentials
+        return credentials
+    }
+
+    private func shouldRefreshCredentials(expiration: Date?) -> Bool {
+        guard let expiration = expiration else {
```

Comment:
nit: you can skip the assignment.
```suggestion
        guard let expiration else {
```

Reply:
Changed as suggested.